### PR TITLE
Add Diffblue Deeptest tests to increase line coverage

### PR DESCRIPTION
**Pull request by petersear**
This is another trial-run of a PR, against the Diffblue fork, to allow us to review the tests before submission of the actual PR. The tests included were identified by test-validator and are those remaining after rejecting any which failed to compile. Suggestions and reasons for tests to leave out are welcome, as are any opinions on what is good about some of these tests. Including everyone in the CX team so we can all see the discussion.

(The following text and the heading of this "PR" are from the commit message in the format discussed with Peter and Paul recently)

The classes covered by these tests are:

FieldNamingPolicy
LazilyParsedNumber
TypeAdapters
ISO8601Utils

In order to ensure these tests run successfully, a selection of dependencies have been added to the POM file (Powermock, Mockito & Diffblue's own Deeptest Utils)

### CODE REVIEW

File: gson/src/test/java/com/google/gson/FieldNamingPolicyTestCase.java [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+   *     - conditional line 150 branch to line 153
+   */
+
+  @Test
+  public void separateCamelCaseInputNotNullNotNullOutputNotNull()
+      throws NoSuchMethodException, IllegalAccessException, InvocationTargetException {
+
+    // Arrange
+    final String name = "!";
+    final String separator = "$$$$$$$";
```
Comment (peterschrammel) [inline gson/src/test/java/com/google/gson/FieldNamingPolicyTestCase.java:33]: Heuristics could possibly help us: something called 'separator' should be a single character, preferably ',', ';', etc.
Can you create a ticket for this please?

File: gson/src/test/java/com/google/gson/internal/LazilyParsedNumberTestCase.java [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+   *  - conditional line 92 branch to line 92
+   */
+
+  @Test
+  public void equalsInputNotNullOutputFalse() throws InvocationTargetException {
+
+    // Arrange
+    final LazilyParsedNumber objectUnderTest =
+        ((LazilyParsedNumber)Reflector.getInstance("com.google.gson.internal.LazilyParsedNumber"));
+    Reflector.setField(objectUnderTest, "value", "\"");
```
Comment (peterschrammel) [inline gson/src/test/java/com/google/gson/internal/LazilyParsedNumberTestCase.java:32]: Not sure what LazilyParsedNumber is exactly doing, but 'Number' would indicate that 'value' should be a number, preferably. Please add this to the heuristics ticket.

File: gson/src/test/java/com/google/gson/internal/bind/TypeAdaptersTestCase.java [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    final Class sub = null;
+    final TypeAdapter typeAdapter = null;
+
+    // Act
+    final TypeAdapterFactory retval =
+        TypeAdapters.newFactoryForMultipleTypes(base, sub, typeAdapter);
+
+    // Assert result
+    Assert.assertNotNull(retval);
+    Assert.assertEquals(null, Reflector.getInstanceField(retval, "val$typeAdapter"));
```
Comment (peterschrammel) [inline gson/src/test/java/com/google/gson/internal/bind/TypeAdaptersTestCase.java:36]: This is not super-useful, unless there is also a test that checks that non-null values are correctly initialised too.

File: gson/src/test/java/com/google/gson/internal/bind/util/ISO8601UtilsTestCase.java [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+   *
+   */
+
+  @Test
+  public void indexOfNonDigitInputNotNullZeroOutputZero2()
+      throws NoSuchMethodException, IllegalAccessException, InvocationTargetException {
+
+    // Arrange
+    final String string = "\"`((((((((";
+    final int offset = 0;
```
Comment (peterschrammel) [inline gson/src/test/java/com/google/gson/internal/bind/util/ISO8601UtilsTestCase.java:97]: Is there no test for non-zero or negative offset?

**Review by peterschrammel — APPROVED** [review]
Looks quite good. Comments concern possible future features.

File: gson/src/test/java/com/google/gson/FieldNamingPolicyTestCase.java (hunk 1 above)
Comment (petersear) [reply gson/src/test/java/com/google/gson/FieldNamingPolicyTestCase.java:33]: Will do.

File: gson/src/test/java/com/google/gson/internal/LazilyParsedNumberTestCase.java (hunk 2 above)
Comment (petersear) [reply gson/src/test/java/com/google/gson/internal/LazilyParsedNumberTestCase.java:32]: OK

File: gson/src/test/java/com/google/gson/internal/bind/TypeAdaptersTestCase.java (hunk 3 above)
Comment (petersear) [reply gson/src/test/java/com/google/gson/internal/bind/TypeAdaptersTestCase.java:36]: I don't see an equivalent test for non-null values.

File: gson/src/test/java/com/google/gson/internal/bind/util/ISO8601UtilsTestCase.java (hunk 4 above)
Comment (petersear) [reply gson/src/test/java/com/google/gson/internal/bind/util/ISO8601UtilsTestCase.java:97]: There are no tests of this method with a non-zero offset.

File: gson/src/test/java/com/google/gson/FieldNamingPolicyTestCase.java (hunk 1 above)
Comment (rjmunro) [reply gson/src/test/java/com/google/gson/FieldNamingPolicyTestCase.java:33]: I'm not convinced. Errors can appear when something like a separator is something unlike what you'd expect it to be. This is the kind of thing we should be good at - what happens if you have a string where the separator is longer than the input etc.

Comment (peterschrammel) [reply gson/src/test/java/com/google/gson/FieldNamingPolicyTestCase.java:33]: Yes, the point is that we shouldn't only test the corner cases, but also the 'normal' case.

**Comment by petersear** [comment]
PR was submitted against the actual repository.
Closing...
